### PR TITLE
Spotinst - adding event subscription module

### PR DESCRIPTION
**Pull request by jeffnoehren**
##### SUMMARY
Added new module for Spotinst Event Subscription 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
Spotinst

##### ADDITIONAL INFORMATION
Added new module to interact with the Spotinst event subscription API

```paste below
result:
    type: str
    sample: sis-e62dfd0f
    returned: success
    description: Created Subscription successfully
```

### CODE REVIEW

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+module: spotinst_event_subscription
+version_added: 2.8
+short_description: Create, update or delete Spotinst Ocean
+author: Spotinst (@jeffnoehren)
+description:
+  - Can create, update, or delete Spotinst Ocean
+    You will have to have a credentials file in this location - <home>/.spotinst/credentials
+    The credentials file must contain a row that looks like this
+    token = <YOUR TOKEN>
+    Full documentation available at https://help.spotinst.com/hc/en-us/articles/115003530285-Ansible-
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]:
```suggestion
    Full documentation available at U(https://help.spotinst.com/hc/en-us/articles/115003530285-Ansible-)
```

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+requirements:
+  - python >= 2.7
+  - spotinst_sdk >= 1.0.44
+options:
+
+  id:
+    description:
+      - (String) Parameters used for Updating or Deleting subscription.
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]:
```suggestion
      - Parameters used for Updating or Deleting subscription.
    type: str
```

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    has_changed = True
+
+    return subscription_id, message, has_changed
+# endregion
+
+
+def main():
+    fields = dict(
+        account_id=dict(type='str'),
+        token=dict(type='str'),
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: You can use `env_fallback` like 

```python
from ansible.module_utils.basic import env_fallback
...
token=dict(type='str', fallback=(env_fallback, ['SPOTINST_TOKEN'])),

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    message = 'Deleted subscription successfully'
+    has_changed = True
+
+    return subscription_id, message, has_changed
+# endregion
+
+
+def main():
+    fields = dict(
+        account_id=dict(type='str'),
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]:
```suggestion
        account_id=dict(type='str', fallback=(env_fallback, ['SPOTINST_ACCOUNT_ID', 'ACCOUNT']),
```

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+    token = module.params.get('token')
+    if not token:
+        token = os.environ.get('SPOTINST_TOKEN')
+    if not token:
+        token = creds_file_loaded_vars.get("token")
+
+    account = module.params.get('account_id')
+    if not account:
+        account = os.environ.get('SPOTINST_ACCOUNT_ID') or os.environ.get('ACCOUNT')
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: No need for this if you are using `env_fallback` mentioned in below comment.

**Review by Akasurde — CHANGES_REQUESTED** [review]
Could you please add a unit test for this change ? Thanks.

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py (hunk 5 above)
Comment (jeffnoehren) [reply lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: added in!

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py (hunk 4 above)
Comment (jeffnoehren) [reply lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: added

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py (hunk 3 above)
Comment (jeffnoehren) [reply lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: added

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py (hunk 2 above)
Comment (jeffnoehren) [reply lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: added

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py (hunk 1 above)
Comment (jeffnoehren) [reply lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py]: added

**Comment by jeffnoehren** [comment]
@Akasurde Hey I added in the changes you suggested as well as a unit test for this module! Also I have three other PR's open right now that I went through and updated the same changes and added unit tests for those three other modules. Please let me know if there is anything else that is required to move forward! 

https://github.com/ansible/ansible/pull/51517
https://github.com/ansible/ansible/pull/51516
https://github.com/ansible/ansible/pull/51514

File: changelogs/fragments/spotinst-add-event-subscription-module.yaml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
```
Comment (s-hertel) [inline changelogs/fragments/spotinst-add-event-subscription-module.yaml:1]: New modules don't need a changelog fragment as they're automatically generated.

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+      - create update or delete
+
+  resource_id:
+    description:
+      - Resource that the subscription will be on
+    type: str
+
+  protocol:
+    description:
+      - (String) Type of desired protocol
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py:63]:
```suggestion
      - Type of desired protocol
    type: str
```

File: lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+  event_format:
+    description:
+      - Event body to be sent to endpoint
+    type: str
+"""
+EXAMPLES = """
+#In this basic example, we create an event subscription
+
+- hosts: localhost
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/spotinst/spotinst_event_subscription.py:83]: Replace line 81 to 97 - 

with 
```yaml
    - name: Create an event subscription
      spotinst_event_subscription:
        account_id:
        token:
        state: present
        id: sis-e62dfd0f
        resource_id: sig-992a78db
        protocol: web
        endpoint: https://webhook.com
        event_type: GROUP_UPDATED
        event_format: { "subject" : "%s", "message" : "%s" }
      register: result
```

**Comment by Akasurde** [comment]
Ping me once you are done with the change. I will take a look into other PRs you mentioned.

**Pull request closed without merging**